ff/src/fields/field_hashers/expander/mod.rs

```diff
@@ -2,98 +2,116 @@
+pub struct DST(arrayvec::ArrayVec<u8,MAX_DST_LENGTH>);
+
+impl DST {
+    pub fn new_xmd<H: FixedOutputReset+Default>(dst: &[u8]) -> DST {
+        DST(if dst.len() > MAX_DST_LENGTH {
+            let mut long = H::default();
+            long.update(&LONG_DST_PREFIX[..]);
+            long.update(&dst);
+            ArrayVec::try_from( long.finalize_fixed().as_ref() ).unwrap()
+        } else {
+            ArrayVec::try_from(dst).unwrap()
+        })
+    }
+
+    pub fn new_xof<H: ExtendableOutput+Default>(dst: &[u8], k: usize) -> DST {
+        DST(if dst.len() > MAX_DST_LENGTH {
+            let mut long = H::default();
+            long.update(&LONG_DST_PREFIX[..]);
+            long.update(&dst);
+
+            let mut new_dst = [0u8; MAX_DST_LENGTH];
+            let new_dst = &mut new_dst[0..((2 * k + 7) >> 3)];
+            long.finalize_xof_into(new_dst);
+            ArrayVec::try_from( &*new_dst ).unwrap()
+        } else {
+            ArrayVec::try_from(dst).unwrap()
+        })
+    }
 
-pub(super) struct ExpanderXof<T: Update + Clone + ExtendableOutput> {
-    pub(super) xofer: T,
+    pub fn update<H: Update>(&self, h: &mut H) {
+        h.update(self.0.as_ref());
+        // I2OSP(len,1) https://www.rfc-editor.org/rfc/rfc8017.txt
+        h.update(&[self.0.len() as u8]);
+    }
+}
+
+
+pub(super) struct ExpanderXof<H: ExtendableOutput + Clone + Default> {
+    pub(super) xofer: H,
     pub(super) dst: Vec<u8>,
     pub(super) k: usize,
 }
 
-impl<T: Update + Clone + ExtendableOutput> Expander for ExpanderXof<T> {
-    fn construct_dst_prime(&self) -> Vec<u8> {
-        let mut dst_prime = if self.dst.len() > MAX_DST_LENGTH {
-            let mut xofer = self.xofer.clone();
-            xofer.update(&LONG_DST_PREFIX.clone());
-            xofer.update(&self.dst);
-            xofer.finalize_boxed((2 * self.k + 7) >> 3).to_vec()
-        } else {
-            self.dst.clone()
-        };
-        dst_prime.push(dst_prime.len() as u8);
-        dst_prime
-    }
+impl<H: ExtendableOutput + Clone + Default> Expander for ExpanderXof<H> {
     fn expand(&self, msg: &[u8], n: usize) -> Vec<u8> {
-        let dst_prime = self.construct_dst_prime();
-        let lib_str = &[((n >> 8) & 0xFF) as u8, (n & 0xFF) as u8];
-
-        let mut xofer = self.xofer.clone();
+        let mut xofer = H::default();
         xofer.update(msg);
-        xofer.update(lib_str);
-        xofer.update(&dst_prime);
-        xofer.finalize_boxed(n).to_vec()
+
+        // I2OSP(len,2) https://www.rfc-editor.org/rfc/rfc8017.txt
+        let lib_str = (n as u16).to_be_bytes();
+        xofer.update(&lib_str);
+
+        DST::new_xof::<H>(self.dst.as_ref(), self.k).update(&mut xofer);
+        xofer.finalize_boxed(n).into_vec()
     }
 }
 
-pub(super) struct ExpanderXmd<T: DynDigest + Clone> {
-    pub(super) hasher: T,
+pub(super) struct ExpanderXmd<H: FixedOutputReset + Default + Clone> {
+    pub(super) hasher: H,
```

Comment:
CI complains that `hasher` is never used/read - I think it refers to its `(super)` visibility, it can probably stay private.

Reply:
We'll change this in a later PR but we've not changed this line yet, so the CI is wrong here.
https://github.com/arkworks-rs/algebra/pull/678/files#diff-ec6c8867ccebc8916382b7a4c0d910966930416987c40d56011dcf45fb88c872R54

This is a common incorrect warning.

Reply:
Hm given that we can't merge without it, what's an acceptable change for right now to get this PR merged?

Reply:
Maybe an `#[allow]` temporarily, that will be removed in a future PR?

Reply:
Oops!  It's fine as `PhantomData`.  This churns later too, but no issue here.  I'll rebase and push the fix.

Reply:
ok fixed now.

Reply:
Alright minor hiccup with tests not being run locally, but now fixed I think.